spanner/api/Program.cs

```diff
@@ -0,0 +1,989 @@
+                    + $"AND AlbumTitle < '{endTitle}'");
+                using (var reader = await cmd.ExecuteReaderAsync())
+                {
+                    while (await reader.ReadAsync())
+                    {
+                        Console.WriteLine($"AlbumId : {reader["AlbumId"]} "
+                        + $"AlbumTitle : {reader["AlbumTitle"]} "
+                        + $"MarketingBudget : {reader["MarketingBudget"]}");
+                    }
+                }
+            }
+            return;
+            // [END read_data_with_storing_index]
+        }
+
+        public static async Task QueryDataWithTransactionAsync(
+    string projectId, string instanceId, string databaseId)
+        {
+            // [START read_only_transaction]
+            string connectionString =
+            $"Data Source=projects/{projectId}/instances/{instanceId}"
+            + $"/databases/{databaseId}";
+            // Gets a transaction object that captures the database state
+            // at a specific point in time.
+            using (TransactionScope scope = new TransactionScope(
+                TransactionScopeAsyncFlowOption.Enabled))
+            {
+                // Create connection to Cloud Spanner.
+                var connection = new SpannerConnection(
+                    connectionString);
+                // Open connection as read only within the scope 
+                // of the current transaction.
+                connection.OpenAsReadOnly();
+                using (connection)
+                {
+                    var cmd = connection.CreateSelectCommand(
+                        "SELECT SingerId, AlbumId, AlbumTitle FROM Albums");
+                    // Read #1.
+                    using (var reader = await cmd.ExecuteReaderAsync())
+                    {
+                        while (await reader.ReadAsync())
+                        {
+                            Console.WriteLine(
+                                $"SingerId : {reader["SingerId"]}"
+                                + $" AlbumId : {reader["AlbumId"]}"
+                                + $" AlbumTitle : {reader["AlbumTitle"]}");
+                        }
+                    }
+                    // Read #2. Even if changes occur in-between the reads, 
+                    // the transaction ensures that Read #1 and Read #2 
+                    // return the same data.
+                    using (var reader = await cmd.ExecuteReaderAsync())
+                    {
+                        while (await reader.ReadAsync())
+                        {
+                            Console.WriteLine(
+                                $"SingerId : {reader["SingerId"]}"
+                                + $" AlbumId : {reader["AlbumId"]}"
+                                + $" AlbumTitle : {reader["AlbumTitle"]}");
+                        }
+                    }
+                }
+                //scope.Complete();
+                return;

```

Comment:
transient errors (like the service being unavailable) can also affect reads and topaz should be used here for retries as well.